test/blackbox-tests/test-cases/rule-target-inferrence/run.t

```diff
@@ -0,0 +1,22 @@
+When a rule target can be be inferred from the rule action, the target or targets field can be omitted.
+This works with the short form of the rule stanza:
+
+  $ dune build @infer --root short-form
+  Entering directory 'short-form'
+  It worked!
+
+But should work with the long form as well:
+
+  $ dune build @infer --root long-form
+  Entering directory 'long-form'
+  It should work as well!
+
+When the targets cannot be inferred from the action, an helpful error message is displayed:
+
+  $ dune build --root cannot-infer
+  Entering directory 'cannot-infer'
+  File "dune", line 1, characters 0-34:
+  1 | (rule (action (echo "something")))
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+  Error: Rule has no targets specified
```

Comment:
@rgrinberg, FWIW, I don't think this is "cannot be inferred", this is "can be inferred to be empty", which is a different situation. If the intention is to make something opaque to dune, you *do* need an external command. (I'm not saying one test is more important or better than the other, just that they test different things)

Reply:
Agreed. Did you mean to mention @NathanReb however? I'd also prefer a message like `dune was not able to infer any targets and none were manually specified`.

Reply:
I meant to mention you because you asked Nathan to use echo. I don't think "none were manually specified" is useful in this situation because that's misleading: specifying targets won't help.

Reply:
Yeah I added the test case to showcase that the error message could be improved. I'll add a different test for the no-target case and use `bash` in this one again.

I'll take a look at how to improve the error message here!

Reply:
Ah I see. So indeed we should have 2 messages:

* No targets were specified
* Dune did not infer any targets

@NathanReb would you like to improve dune to distinguish between these?

Reply:
@NathanReb there's still no need to use bash btw. Use `system`.

Reply:
I added the test case but from a very quick look it seems like it's going to take a tiny bit more effort to distinguish those two error cases so feel free to merge as it if you need to.

I'll work on that separately!